### PR TITLE
Refactor the core/data store to be independent from the registry

### PR DESCRIPTION
**Pull request by youknowriad**
Related #14369 

There's a lot in this PR but I don't know if it's breakable into smaller items. The main ideas here:

 - Bake-in the "core/data" store as a reducer embedded in any namespace store
 - Build a "virtual" store added to all registries but proxify its selectors/actions per store
 - Bake the controls plugin in the namespace stores (it wasn't possible to keep it separate and bake "core/data")

This is not ready as I didn't check the tests yet.

### CODE REVIEW

File: packages/data/src/registry.js
Comment (nerrad) [inline packages/data/src/registry.js]: As an aside, can we switch the variable name from `reducerKey` to `storeKey`?  It could help with #14150.  I think this pull would be an appropriate place to do it because of the refactoring taking place.

**Review by nerrad — COMMENTED** [review]
I like the direction of this.

> Build a "virtual" store added to all registries but proxify its selectors/actions per store

I'm assuming this refers to the `core/data` generic store instantiated for each created registry.  If so, I have to admit the terminology gets me tripped up sometimes because a registry can have multiple _stores_ registered on it, so the above sentence reads better to me as "proxify its selectors/actions per _registry_".  

> Bake the controls plugin in the namespace stores (it wasn't possible to keep it separate and bake "core/data")

I think this seems sensible given the functionality of controls.

File: packages/data/src/namespace-store/index.js
Comment (youknowriad) [reply packages/data/src/namespace-store/index.js]: there are multiple ways to provide "fulfill" (this existed before, I just moved it because the implementation changed a little bit).

you can write 

`resolver = { fulfill: () => { }, isFullfilled: () => { } }`

or


```
resolver = function() {}
resolver.isFullfilled = () => {}
```

Comment (youknowriad) [reply packages/data/src/namespace-store/index.js]: this code normalizes all these approaches into one, that way the following code don't need "ifs"

File: packages/data/src/store/index.js
Comment (youknowriad) [reply packages/data/src/store/index.js]: yes

File: packages/data/src/store/index.js
Comment (youknowriad) [reply packages/data/src/store/index.js]: Because there's no store stored in this store that need to retrigger selectors, if a change in the resolution state happens, the corresponding store would have already triggered a "subscribe" call.

File: packages/data/src/registry.js
Comment (youknowriad) [reply packages/data/src/registry.js]: I think it should be "storeName" but given how widespread this change could be, I think it's better in its own PR.

**Comment by nerrad** [comment]
With the controls plugin being baked in, I wonder if - in a separate pull after this lands - if we should also bake in some common controls often used.  Things like:

- apiFetch
- select
- dispatch
- resolveSelect
- resolveDispatch (for dispatches that return a value)
- subscribe

**Comment by youknowriad** [comment]
@nerrad or maybe a dedicated `@wordpress/data-controls`

**Comment by nerrad** [comment]
> or maybe a dedicated @wordpress/data-controls

Oh ya, even better.  That's something that can be done irrespective of this landing.  I might give a go at it sometime soon.

File: packages/data/src/namespace-store/index.js
Comment (aduth) [inline packages/data/src/namespace-store/index.js]: Nit: Good candidate for destructure `const { metadata }`

File: packages/data/src/plugins/persistence/index.js
Comment (aduth) [inline packages/data/src/plugins/persistence/index.js]: Can you describe what's changing in this file? Is it related to the prior comment note of "some modules monkey-patching the store object" ?

File: packages/data/src/registry.js
Comment (aduth) [inline packages/data/src/registry.js]: Nit: It's a good opportunity to stop using `Array#map` wrongly (vs. the more appropriate `Array#forEach`)

File: packages/data/src/store/index.js
Comment (aduth) [reply packages/data/src/store/index.js]: > Because there's no store stored in this store that need to retrigger selectors, if a change in the resolution state happens, the corresponding store would have already triggered a "subscribe" call.

Can we add this as an inline code comment for the benefit of the next maintainer?

File: packages/data/src/namespace-store/index.js
Comment (aduth) [inline packages/data/src/namespace-store/index.js]: An initial thought occurred to me that there's risk for conflict between these metadata actions and a store's own. On further consideration, I don't see it likely to occur, or at least it should be known that these "baseline" actions/selectors exist for every store. Which leads me to wonder: How do we document this?

File: packages/data/src/namespace-store/index.js
Comment (aduth) [inline packages/data/src/namespace-store/index.js]: This logic occurs for every selector call? In addition to the overhead of [`mapSelectors`](https://github.com/WordPress/gutenberg/blob/8243a1c9f6561c2dcbd0874349e044a5467c9321/packages/data/src/namespace-store.js#L111)? Should expect a negative performance impact?

Comment (aduth) [reply packages/data/src/namespace-store/index.js]: I wonder if it's not unreasonable to consider/implement what we have currently as `metadata` and `root` as in-fact two separate stores, so that we don't have to override the selector call to reach into state to provide the corrected "root" state.

**Review by aduth — COMMENTED** [review]
I personally don't have an issue with making `controls` plugin default behavior. The documentation should be updated accordingly.

File: packages/data/src/plugins/persistence/index.js
Comment (youknowriad) [reply packages/data/src/plugins/persistence/index.js]: exactly, I made this change prior to the monkey-patching fix. So we could revert it now I think, but ultimately I think this is a good change, so I'm thinking of keeping it now.

File: packages/data/src/namespace-store/index.js
Comment (youknowriad) [reply packages/data/src/namespace-store/index.js]: yes, I assume these can't be auto-documented. I'll see if I can add a note in the data module README (where we document `registerStore`)

File: packages/data/src/namespace-store/index.js
Comment (youknowriad) [reply packages/data/src/namespace-store/index.js]: to your second point, in theory, a store could be just a scalar value which would break this proposal.

to your first point, this only happens when registering the store which I guess is fine. (only run once)

Comment (aduth) [reply packages/data/src/namespace-store/index.js]: > to your first point, this only happens when registering the store which I guess is fine. (only run once)

Subsequent calls would still be subject to going through the `( state, ...args ) => selector( state.root, ...args )` proxying? May be small overhead; curious though given the reality of many thousands of selector calls during typical use.

File: packages/data/src/namespace-store/index.js
Comment (aduth) [reply packages/data/src/namespace-store/index.js]: > yes, I assume these can't be auto-documented.

I'm not so overly concerned about the selectors themselves (though it would be nice), more to just general awareness that this is a behavior to expect, the fact that resolution behavior is included in every store by default.

File: packages/data/src/store/index.js
Comment (nerrad) [reply packages/data/src/store/index.js]: Would a more accurate rendition include "Because there's no store stored in this _registry_..."?

File: packages/data/src/registry.js
Comment (nerrad) [reply packages/data/src/registry.js]: ooo good catch there.

File: packages/data/src/namespace-store/index.js
Comment (nerrad) [reply packages/data/src/namespace-store/index.js]: There's potential side-effects here of consuming code being able to "override" default metadata selectors and actions.  I almost wonder if there should be some enforcement here: i.e. checking the incoming actions/selectors and ensuring there are no collisions with the corresponding metadata functions?

Comment (youknowriad) [reply packages/data/src/namespace-store/index.js]: If we're not confident exposing these selectors and actions yet, we can easily add `__internal` or `__experimental` for now, what do you think?

**Comment by youknowriad** [comment]
Anything that you consider a blocker with this. I'd like to land this to unblock a few other PRs

File: packages/data/src/namespace-store/index.js
Comment (nerrad) [reply packages/data/src/namespace-store/index.js]: I'm less concerned with exposing them (I think they _should_ be exposed?) and more concerned with potential conflicts.  

I do think it'd be better (at least initially) to do this:

```
// for actions
const actions = mapActions( {
    ...options.actions,
    ...metadataActions,
} );
// and something similar with selectors reversing the order
```

That way the metadataActions and metadataSelectors take precedence.  Of course you may _intentionally_ have them this way so they can be overridden by the incoming store config but if not, I think this is safer.

**Review by nerrad — APPROVED** [review]
Not a blocker, but I do think (as slim as it is) there's potential for conflict with the resolution selectors (less so actions) (see my earlier comment).

**Comment by aduth** [comment]
Curious why, in Redux DevTools, I see the _selectors_ shown for metadata state. 🤔 

image

**Review by aduth — APPROVED** [review]
I wasn't able to easily test this against my [aforementioned use-case](https://github.com/WordPress/gutenberg/pull/14369#discussion_r268179596) since it requires a combination of this and #14369, but conceptually this seems like a good solution to the general problem.

Still curious about what I'm seeing with Redux DevTools, but preemptively approving.

**Comment by nerrad** [comment]
> Curious why, in Redux DevTools, I see the selectors shown for metadata state

Isn't that because resolution state is stored with the selector as the key (i.e no longer need to store indexed by store name)

**Comment by aduth** [comment]
> > Curious why, in Redux DevTools, I see the selectors shown for metadata state
> 
> Isn't that because resolution state is stored with the selector as the key (i.e no longer need to store indexed by store name)

Ah! I think you're right. I may still think it's a bit unnatural to see those as direct descendents of a "metadata" key. The reducer is called `isResolved`, but because it's the default this key isn't shown. Makes me wonder if we ought either rename `metadata` or have `isResolved` as a proper key of the metadata state.

Small naming quibble, I'm not overly concerned about it now that it's been explained.

**Comment by youknowriad** [comment]
@aduth It does seem like we could add a `metadata.resolution` or something like that for this state. If we need more "metadata" for data stores, we can refactor this.

File: packages/data/src/namespace-store/index.js
Comment (aduth) [inline packages/data/src/namespace-store/index.js]: I seem to be running into an issue where this doesn't actually wait for the fulfillment to complete before marking the resolution as finished:

Screen Shot 2019-03-29 at 11 58 03 AM

_(Note in last three actions, I'd have expected "RECEIVE_ITEMS" to have occurred prior to "FINISH_RESOLUTION")_

I tried to consider how to reproduce this outside my branch, but since we [wait for most all data to be available](https://github.com/WordPress/gutenberg/blob/39171ca106b288b6ae1af06563a013a52b84ed2b/packages/edit-post/src/editor.js#L77) before rendering anything, it's hard to recreate even when adding a filter like `add_filter( 'block_editor_preload_paths', '__return_empty_array' );`. Which has me wondering if it could in-fact be an issue which had already existed prior to this change, and only became evident when performing resolution for data which hadn't been initially requested in the display of the editor (in my case, [waiting to resolve `getPostType( 'wp_block' )`](https://github.com/WordPress/gutenberg/blob/39171ca106b288b6ae1af06563a013a52b84ed2b/packages/editor/src/store/actions.js#L309-L313) when saving a reusable block).

I'll try to clean up my local working branch to push later this afternoon, but raising now in case anything comes to mind.

Comment (aduth) [reply packages/data/src/namespace-store/index.js]: I found the issue. I'll have a pull request shortly.

**Pull request merged**
